### ISSUE TITLE
Feature request: .Ensure(Observed/Subscribed)On(IScheduler)

### ISSUE THREAD

**Issue by gentledepp** [dotnet/reactive#1977]
When using RX, it is sometimes tedious to fine-tune which thread a subscription or observer is handled on.

This is especially painful when using RX with UI frameworks such as Xamarin.Forms.

Adding `ObserveOn` explicitly every time is the way to go for now, but this is said to come with a performance overhead. (Yeah - I _did_ read _You, I and reactive UI_...

So it would be nice to have a `EnsureObservedOn(IScheduler)` which checks, _if_ the current scheduler is already the one in question and _only if not_ re-schedules the callback to the targeted scheduler.

However, now that I am thinking of it, I appears to me that the only case we ever need this is when we want to ensure that some callback is called on the UI Thread.
So maybe this is actually an issue for Reactive.UI?

@idg10 - this is another minor pain point. Maybe you can figure something out?

**Comment by idg10** [comment]
I want to make sure I've understood what you're looking for. I believe you want to avoid the cost of going through a dispatcher or similar UI thread message pump mechanism in cases where we are already on the UI thread?

I don't think it can work quite how you've proposed:

>checks, if the current scheduler is already the one

because there's no notion of "the current scheduler" in Rx.

There might be several. It's entirely possible to be in a context where you are on a WPF dispatcher thread because `ObserveOn` was used, but also to be inside, say, a `CurrentThreadScheduler` because something downstream of that `ObserveOn` did some work that used that. I think this would have that effect:

```cs
source
    .ObserveOn(DispatcherScheduler.Current)
    .SelectMany(_ => Observable.Range(1, 10))
    .Subscribe(i => Console.WriteLine("I have *multiple* current schedulers!"));
```

That overload of `Observable.Range` uses `CurrentThreadScheduler.Instance` to perform its iteration, so the callback passed to subscribe is actually running in the context of two different schedulers. So if there were to be some concept of "current scheduler" in Rx, it would need to be a stack of current schedulers.

In any case, we don't have any sort of tracking that would enable us to learn anything about which schedulers are current.

There's a further complication: `DispatcherScheduler.Current` actually returns a new instance every time you fetch it. So even if we did have a mechanism for asking "am I running in the context of this scheduler?" it wouldn't help this example, because there can be any number of different dispatcher schedulers for a single `Dispatcher`, so the answer might well be "no" even when you are already on the right dispatcher.

(Also, I'd be a bit reluctant to officially bless coding styles in which you don't know what scheduler you're running on. It's generally best to have a top-level plan for your threading and concurrency requirements. Trying to handle scheduling issues locally is often a mistake, and this proposal certainly seems to be about enabling some code to try and do the right thing in cases where it doesn't have the necessary global knowledge to know whether the `ObserveOn` is really required. My usual recommendation would be "make sure you do know" because it is hard to write correct code when you don't know.)

So for those reasons I don't think we can do it in the way you suggest.

But what might work is something more scheduler-specific. So perhaps an `EnsureObservedOnDispatcher` that is explicitly about WPF. It would ask not "are we on this scheduler?" but "are we already running in the context of the right dispatcher?"

Even then, I'm a bit reluctant because it encourages a coding practice that, in my experience at least, tends to lead to bugs. Concurrency is an application-level concern and is, in my view, best handled holistically. And this is very much a non-holistic approach.

But if my suggested slightly different approach would work for you, I'd certainly be open to looking how it would work in practice. But I think I'd like to see some more specific detailed scenarios in which you (or anyone else who would like this feature) are looking to use this, because I'd need to be convinced that there really are scenarios in which a more structured, holistic approach to concurrency wouldn't be a better way to address the problem.

**Comment by gentledepp** [comment]
Thank you for the detailed response! (And sorry for my late one - I was on holiday)

Well, I thought about this a bit more. The _only_ case where this `EnsureObservedOn...` is required is when we need to make sure, that the callback is done on the UI Thread.
So I guess this is really not a responsibility of RX, but of the UI framework.

In _ReactiveUI_ there _is_ the notion of a  `RxApp.MainThreadScheduler` and a `RxApp.TaskPoolScheduler`
Therefore I guess I should ask there to include a `EnsureObservedOnMainThreadScheduler` extension method.

While I agree that one should always have a strategy for concurrency, the reality is different:
Not every developer _knows_ how to handle concurrency. Especially junior devs do sometimes even struggle with the "task vs thread" topic....

The problem with RX is, that, while being extremely useful _iif_ you know how to use it, it provides tons of pitfalls if you _don't_.
And I am trying to figure out ways to make this technology more approachable.

Thank you for your thoughts!

**Comment by idg10** [comment]
> So I guess this is really not a responsibility of RX, but of the UI framework.

Normally I think of "UI framework" as meaning something like WPF, Windows Forms, or Avalonia.

For that definition, of UI framework, then this is not the responsibility of the UI framework. It is a requirement imposed by the UI framework. The responsibility to meet this requirement lies with application code that uses that framework.

However, if we include things like ReactiveUI as a kind of "UI framework" then that's slightly different. But only slightly. A library such as ReactiveUI can certainly provide tools to help you meet these requirements. But even then, I don't think it _can_ ultimately be entirely the responsibility of ReactiveUI to abide by this rule. (At least, it's not possible given the amount of flexibility ReactiveUI offers.)

The only way that a library or framework really could be wholly responsible for ensuring that you can't break this threading rule would be if it made it impossible to break it. But as far as I'm aware, the only way yet discovered of doing that is to prevent people from directly using multiple threads in the first place. Either you make concurrency completely impossible, or you encapsulate it in some library that uses it internally but which presents a simple sequential API to its consumers. (E.g., PyTorch does exactly that. And on a grander scale that's basically what Spark does too: it makes the computational power of multithreaded, distributed computation accessible to people who might not have any engineering expertise in concurrency of any kind. But it does so by providing a set of abstractions that completely conceal the concurrency. That is a completely valid engineering choice, and is a large part of why Spark has become so successful, but it's not the right thing for Rx. It is most likely the right thing for some libraries that build on Rx, but not for Rx itself.)

`ObserveOn` and similar abstractions make the multithreaded nature of the system explicit. So they are exactly the opposite of what is required if you want to guarantee that all threading rules are always met.


>While I agree that one should always have a strategy for concurrency, the reality is different:
>Not every developer knows how to handle concurrency. Especially junior devs do sometimes even struggle with the "task vs thread" topic....

This is true, but I think it is unrealistic to suppose that these developers will be able to make successful use of multiple threads if that's their current depth of understanding. They basically have three options:

1. learn
2. don't attempt the advanced stuff
3. use some library that operates at a level of abstraction where you can't even see that it's using multithreading

Medical students in their first year lack the skills to perform open heart surgery. This doesn't mean that we need to find a way of letting them do it. Some things should only be tackled once you have developed the necessary level of expertise.

And this is not to say that Rx should not be used by junior developers. It's that if you're building multi-threaded applications, and frameworks that have particular rules around threading, those junior developers will need some help to navigate this. In a lot of cases, the right answer will be for senior developers to establish rules around how the application uses Rx, and junior developers can either work within those constraints, or learn enough to understand when they can bend the rules.

The kinds of thing being proposed in this discussion are inherently about making threading visible, so that rules out 2 and 3 in the particular scenarios we're discussing. A sufficiently opinionated library might be able to build on top of Rx, and make 3 possible while still presenting Rx abstractions to the developer, but I don't think Rx itself is the place for that. It should provide the tools for this sort of thing, but it shouldn't impose the constraints that would be necessary to enforce correctness, because there isn't one right way to do that. (E.g., "no multithreading allowed" is one way to do that, but Rx certainly shouldn't force that on everyone, even though it is the simplest solution, and for an awful lot of scenarios it is the best choice in practice.)

Maybe ReactiveUI could enable 3, but I'm not sure they're trying to impose those kinds of constraints either.

But certainly for Rx, I think the only options we can offer novice developers are 1 or 2. I would not expect any developer who doesn't understand the distinction between a task and a thread to be able to make successful use of a mechanism of the kind you are proposing here.
